Mathlib/Combinatorics/Enumerative/DyckWord.lean

```diff
@@ -196,4 +209,143 @@ lemma two_mul_semilength_eq_length : 2 * p.semilength = p.toList.length := by
+        take_zero, append_nil, count_take_firstReturn_add_one h]
+    · intro j hj
+      rw [take_append_eq_append_take, show j + 1 - p.toList.length = 0 by omega,
+        take_zero, append_nil]
+      exact (count_lt_count_of_lt_firstReturn hj).ne'
+  · rw [length_range, u, length_append]
+    exact Nat.lt_add_right _ (firstReturn_lt_length h)
+
+@[simp]
+lemma firstReturn_nest : p.nest.firstReturn = p.toList.length + 1 := by
+  have u : p.nest.toList = U :: p.toList ++ [D] := rfl
+  rw [firstReturn, findIdx_eq]
+  · simp_rw [get_eq_getElem, getElem_range, u, decide_eq_true_eq]
+    constructor
+    · rw [take_of_length_le (by simp), ← u, p.nest.count_U_eq_count_D]
+    · intro j hj
+      simp_rw [cons_append, take_cons, count_cons, beq_self_eq_true, ite_true,
+        beq_iff_eq, ite_false, take_append_eq_append_take,
+        show j - p.toList.length = 0 by omega, take_zero, append_nil]
+      have := p.count_D_le_count_U j
+      omega
+  · simp_rw [length_range, u, length_append, length_cons]
+    exact Nat.lt_add_one _
+
+lemma nest_add_firstReturn : (p.nest + q).firstReturn = p.toList.length + 1 := by
+  rw [firstReturn_add, firstReturn_nest, ite_eq_right_iff, ← toList_eq_nil]; intro; contradiction
+
+variable (p) in
+/-- The left part of the Dyck word decomposition,
+inside the `U,D` pair that `firstReturn` refers to. `insidePart 0 = 0`. -/
+def insidePart : DyckWord :=
+  if h : p = 0 then 0 else
+  (p.take (p.firstReturn + 1) (count_take_firstReturn_add_one h)).denest
+    (by rw [← toList_ne_nil, take]; simpa using toList_ne_nil.mpr h)
+    (fun i lb ub ↦ by
+      simp only [take, length_take, lt_min_iff] at ub ⊢
+      replace ub := ub.1
+      rw [take_take, min_eq_left ub.le]
+      rw [show i = i - 1 + 1 by omega] at ub ⊢
+      rw [Nat.add_lt_add_iff_right] at ub
+      exact count_lt_count_of_lt_firstReturn ub)
+
+variable (p) in
+/-- The right part of the Dyck word decomposition,
+outside the `U,D` pair that `firstReturn` refers to. `outsidePart 0 = 0`. -/
+def outsidePart : DyckWord :=
+  if h : p = 0 then 0 else p.drop (p.firstReturn + 1) (count_take_firstReturn_add_one h)
+
+@[simp] lemma insidePart_zero : insidePart 0 = 0 := by simp [insidePart]
+@[simp] lemma outsidePart_zero : outsidePart 0 = 0 := by simp [outsidePart]
+
+include h in
+@[simp]
+theorem insidePart_nest_add_outsidePart : p.insidePart.nest + p.outsidePart = p := by
+  simp_rw [insidePart, outsidePart, h, dite_false, denest_nest, DyckWord.ext_iff]
+  apply take_append_drop
+
+include h in
+theorem semilength_insidePart_lt : p.insidePart.semilength < p.semilength := by
+  rw [← Nat.mul_lt_mul_left zero_lt_two]
+  simp_rw [two_mul_semilength_eq_length]
+  nth_rw 2 [← insidePart_nest_add_outsidePart h]
+  change _ < (U :: p.insidePart.toList ++ [D] ++ p.outsidePart.toList).length
+  simp_rw [length_append, length_singleton, length_cons]; omega
+
+include h in
+theorem semilength_outsidePart_lt : p.outsidePart.semilength < p.semilength := by
```

Comment:
Can you deduce both lemmas from `p.insidePart.semilength + p.outsidePart.semilength + 1 = p.semilength`, which you itself can deduce from `p.insidePart.length + p.outsidePart.length + 2 = p.length`?

Reply:
All done.